### ISSUE TITLE
ROCm: Questioning the `:device=` and `:instance=` order

### ISSUE THREAD

**Issue by bertwesarg** [icl-utk-edu/papi#100]
So event names for ROcm are `rocm:::<name>:device=<device>[:instance=]`. Here is an example event name:

```
--------------------------------------------------------------------------------
| rocm:::TCP_TCP_TA_DATA_STALL_CYCLES:device=0:instance=0                      |
|            TCP stalls TA data interface. Now Windowed.                       |
--------------------------------------------------------------------------------
| rocm:::TCP_TCP_TA_DATA_STALL_CYCLES:device=0:instance=1                      |
|            TCP stalls TA data interface. Now Windowed.                       |
--------------------------------------------------------------------------------
```

In [Score-P](https://score-p.org) we want to make it convenient for the user to just tell the measurement system, which metrics to record, on all devices. Hence, the user should only configure Score-P with the metric name (e.g., `rocm:::TCP_TCP_TA_DATA_STALL_CYCLES`) and will generate the proper event name depending on which device it is running. this is important in an multi-GPU/multi-node experiment, where the user does not know in advance which device is allocated on the node. Hence, the user is only interested in the metric name to record.

But what to do with the _instances_? As the order is fixed, Score-P would need to split the metric name on `:instance=` and insert the `:device=`. This seems very fragile.

Hence I questioning the usefulness of the `:device=` and `:instance=` ordering and propose to change it to `rocm:::<name>[:instance=]:device=<device>`. This way Score-P can always just append `:device=` and it works.

A different approach would be to know which metric has instances at all and how many and also append the `:instance=` to the base metric name. But I doubt there is such API in PAPI available for this.

**Comment by gcongiu** [comment]
The naming format can surely be changed in the ROCm component. I am not sure this is a PAPI concern thought. Score-p users should know what metric and instance to monitor as metrics with different instance number are different events. If the user is not interested in any particular instance there are accumulated metrics that sum up all the instances values and could be used instead. Score-p adding the instance itself sounds like an arbitrary choice to me.

**Comment by bertwesarg** [comment]
No, I was not planning to add the instance number in Score-P. Thats the whole point. The user wants to record instance Y from metric X. There are two devices Z={0,1}. The result must be `rocm:::X:device=Z:instance=Y`. What should the user specify to Score-P? It should not include `:device=Z`, because this will be decide at runtime. But if the user specifies `rocm:::X:device=Z:instance=Y`, Score-P would need to split the name to insert `:device=Z` into the correct position.

> If the user is not interested in any particular instance there are accumulated metrics that sum up all the instances values and could be used instead.

Thanks for this. I think Score-P could then just prohibit `:instance=` event names. This level of detail is probably enough.

**Comment by gcongiu** [comment]
You could prohibit Score-p users to use PAPI native GPU events that need an instance number, but that would be confusing to users for one and also unfair because it limits what they can measure.

I don’t think changing the event name format in PAPI to accomodate for a Score-p introduced behaviour (i.e. allow the user to introduce an ambiguity that is eventually resolved by Score-p itself) would be reasonable. There might be applications using PAPI that rely on that format.

A better solution to your problem might be to introduce wild card events support in PAPI. Something like event_name:device=*:instance=0 could be defined to let the user indicate that he/she wants to measure a certain event name on all GPUs.

**Comment by bertwesarg** [comment]
That would require parsing too, and a more complicated parsing than inserting `:device=x`. If the user specifies `rocm:::event_name` to Score-P, the metric will already be recorded on all nodes and on all devices the individual ranks are accessing. Adding a wildcard like `:device=*` is counter productive here. and worse the user could think he can also specify specific devices in this format, which does not make sense to begin with.

I tend to stick to the insert method then.

**Comment by gcongiu** [comment]
In the wildcard approach Score-P would not need to do any parsing, that was the point of the wildcard. Score-P would just pass the wildcard event to PAPI that would know how to treat it. However, I am not sure this would work for all cases.

**Comment by bertwesarg** [comment]
This is an idea which still would need to be implemented in PAPI, right? If so, it does not help anything on Score-P's side. We have one event set per device, PAPI would need to only add the event for this device to the event set. But how to tell PAPI, that this event set is for device X only?

But again, the user should not care about the `:device=` suffix at all.

Even worse, encoding the location into the event name is also not convenient to the user, when looking at the measurement result. When one wants to look at all event date of all devices, the user would need to select all the events with different `"device=X` suffixes in the GUI, instead of just one without the `:device=` suffix. (or the GUI needs to do some parsing by itself again)

If the order would be reversed, we could avoid all these peculiarities to the user.

**Comment by gcongiu** [comment]
We would need to discuss internally before doing any change to the event name format. I will update the issue once we have done that.

**Comment by gcongiu** [comment]
@bertwesarg, attached is a patch that should allow Score-P users to omit the device id. This means that the ROCm component in PAPI will resolve the event ambiguity by figuring out the currently selected device (through hipGetDevice). This is just an attempt solution, take it with a grain of salt: https://github.com/icl-utk-edu/papi/commit/f305381a30b5f6ee4413aacc91700ab25ae270f0.

The good thing about the attached solution is that it is backward compatible, i.e., if the user decides to pass in the device id everything will work as before. Note that I haven't tested this very extensively. Let me know if you encounter any problem with it.

**Comment by bertwesarg** [comment]
Thanks, this for sure clashes with #75. But the next question is then, if the CUDA component will provide the same convenience? Else Score-P would need to make a per-component decisions whether to add `:device=` or not.

**Comment by gcongiu** [comment]
As I said this is just an attempt. If we go for this solution then yes, all the components should be updated to reflect the same behavior. I am not sure how this would clash with #75. IIRC that is just updating the README to warn the user about the GPU isolation/partitioning mechanism.

**Comment by bertwesarg** [comment]
… and talks explicitly about the ROCm runtime, **not** the HIP runtime, but you call `hipDeviceGet`, which is the HIP runtime

**Comment by gcongiu** [comment]
> … and talks explicitly about the ROCm runtime, **not** the HIP runtime, but you call `hipDeviceGet`, which is the HIP runtime

Fair enough, that can be changed later to use the ROCm runtime function (assuming there is an equivalent call). Functionality should not change.

**Comment by bertwesarg** [comment]
In general, I think this solves my issue, if PAPI takes over the parsing. Thanks.

**Comment by bertwesarg** [comment]
Closed with #123
